Python: introduce NIX_PYTHONPREFIX in order to set site.PREFIXES

adisbladis opened the pull request:
This PR is a story where your valiant hero sets out on a very simple adventure but ends up having to slay dragons, starts questioning his own sanity and finally manages to gain enough knowledge to slay the evil dragon and finally win the proverbial price.

It all started out on sunny spring day with trying to tackle the Nixops plugin infrastructure and make that nice enough to work with.

Our story begins in the shanty town of [NixOps-AWS](https://github.com/nixos/nixops-aws) where [mypy](http://mypy-lang.org/) type checking has not yet been seen.

As our deuteragonist (@grahamc) has made great strides in the capital city of [NixOps](https://github.com/nixos/nixops) our hero wanted to bring this out into the land and let the people rejoice in reliability and a wonderful development experience.

The plugin work itself was straight forward and our hero quickly slayed the first small dragon, at this point things felt good and our hero thought he was going to reach the town of NixOps-AWS very quickly.

But alas! Mypy did not want to go, it said:
`Cannot find implementation or library stub for module named 'nixops'`

Our hero felt a small sliver of life escape from his body. Things were not going to be so easy.

After some frustration our hero discovered there was a [rule of the land of Python](https://www.python.org/dev/peps/pep-0561/) that governed the import of types into the kingdom, more specificaly a very special document (file) called `py.typed`.
Things were looking good.

But no, what the law said did not seem to match reality. How could things be so?

After some frustrating debugging our valiant hero thought to himself "Hmm, I wonder if this is simply a Nix idiosyncrasy", and it turns out indeed it was.
Things that were working in the blessed way of the land of Python (inside a `virtualenv`) were not working the way they were from his home town of Nix (`nix-shell` + `python.withPackages`).

After even more frustrating attempts at reading the mypy documentation and trying to understand how things were supposed to work our hero started questioning his sanity.
This is where things started to get truly interesting.

Our hero started to use a number of powerful weapons, both forged in the land of Python (pdb) & by the mages of UNIX (printf-style-debugging & strace).

After first trying to slay the dragon simply by `strace` and a keen eye our hero did not spot any weak points.
Time to break out a more powerful sword (`pdb`) which also did not divulge any secrets about what was wrong.

Our hero went back to the `strace` output and after a fair bit of thought and analysis a pattern started to emerge. Mypy was looking in the wrong place (i.e. not in in the environment created by `python.withPackages` but in the interpreter store path) and our princess was in another castle!

Our hero went to the pub full of old grumpy men giving out the inner workings of the open source universe (Github) and acquired a copy of Mypy.
He littered the code with print statements & break points.
After a fierce battle full of blood, sweat & tears he ended up in https://github.com/python/mypy/blob/20f7f2dd71c21bde4d3d99f9ab69bf6670c7fa03/mypy/sitepkgs.py and realised that everything came down to the Python `site` module and more specifically https://docs.python.org/3.7/library/site.html#site.getsitepackages which in turn relies on https://docs.python.org/3.7/library/site.html#site.PREFIXES .

Our hero created a copy of the environment created by `python.withPackages` and manually modified it to confirm his findings, and it turned out it was indeed the case.
Our hero had damaged the dragon and it was time for a celebration.

He went out and acquired some mead which he ingested while he typed up his story and waited for the dragon to finally die (the commit caused a mass-rebuild, I had to wait for my repro).

Finally the dragon had been slayed.
In the end all was good in [NixOps-AWS](https://github.com/nixos/nixops-aws)-town and type checks could run. (PR for that incoming tomorrow).

###### Motivation for this change
This is needed in case of `python.buildEnv` to make sure site.PREFIXES
does not only point to the unwrapped executable prefix.

A minimal test case is published at https://github.com/adisbladis/nixpkgs/commit/174bd428da98753e8f8369a0bbe15a58d3873151.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

adisbladis committed:
Python: introduce NIX_PYTHONPREFIX in order to set site.PREFIXES

This is needed in case of `python.buildEnv` to make sure site.PREFIXES
does not only point to the unwrapped executable prefix.

--------------------------------------------------------------------------------

This PR is a story where your valiant hero sets out on a very simple adventure but ends up having to slay dragons, starts questioning his own sanity and finally manages to gain enough knowledge to slay the evil dragon and finally win the proverbial price.

It all started out on sunny spring day with trying to tackle the Nixops plugin infrastructure and make that nice enough to work with.

Our story begins in the shanty town of [NixOps-AWS](https://github.com/nixos/nixops-aws) where [mypy](http://mypy-lang.org/) type checking has not yet been seen.

As our deuteragonist (@grahamc) has made great strides in the capital city of [NixOps](https://github.com/nixos/nixops) our hero wanted to bring this out into the land and let the people rejoice in reliability and a wonderful development experience.

The plugin work itself was straight forward and our hero quickly slayed the first small dragon, at this point things felt good and our hero thought he was going to reach the town of NixOps-AWS very quickly.

But alas! Mypy did not want to go, it said:
`Cannot find implementation or library stub for module named 'nixops'`

Our hero felt a small sliver of life escape from his body. Things were not going to be so easy.

After some frustration our hero discovered there was a [rule of the land of Python](https://www.python.org/dev/peps/pep-0561/) that governed the import of types into the kingdom, more specificaly a very special document (file) called `py.typed`.
Things were looking good.

But no, what the law said did not seem to match reality. How could things be so?

After some frustrating debugging our valiant hero thought to himself "Hmm, I wonder if this is simply a Nix idiosyncrasy", and it turns out indeed it was.
Things that were working in the blessed way of the land of Python (inside a `virtualenv`) were not working the way they were from his home town of Nix (`nix-shell` + `python.withPackages`).

After even more frustrating attempts at reading the mypy documentation and trying to understand how things were supposed to work our hero started questioning his sanity.
This is where things started to get truly interesting.

Our hero started to use a number of powerful weapons, both forged in the land of Python (pdb) & by the mages of UNIX (printf-style-debugging & strace).

After first trying to slay the dragon simply by `strace` and a keen eye our hero did not spot any weak points.
Time to break out a more powerful sword (`pdb`) which also did not divulge any secrets about what was wrong.

Our hero went back to the `strace` output and after a fair bit of thought and analysis a pattern started to emerge. Mypy was looking in the wrong place (i.e. not in in the environment created by `python.withPackages` but in the interpreter store path) and our princess was in another castle!

Our hero went to the pub full of old grumpy men giving out the inner workings of the open source universe (Github) and acquired a copy of Mypy.
He littered the code with print statements & break points.
After a fierce battle full of blood, sweat & tears he ended up in https://github.com/python/mypy/blob/20f7f2dd71c21bde4d3d99f9ab69bf6670c7fa03/mypy/sitepkgs.py and realised that everything came down to the Python `site` module and more specifically https://docs.python.org/3.7/library/site.html#site.getsitepackages which in turn relies on https://docs.python.org/3.7/library/site.html#site.PREFIXES .

Our hero created a copy of the environment created by `python.withPackages` and manually modified it to confirm his findings, and it turned out it was indeed the case.
Our hero had damaged the dragon and it was time for a celebration.

He went out and acquired some mead which he ingested while he typed up his story and waited for the dragon to finally die (the commit caused a mass-rebuild, I had to wait for my repro).

In the end all was good in [NixOps-AWS](https://github.com/nixos/nixops-aws)-town and type checks could run. (PR for that incoming tomorrow).

grahamc reviewed (commented):
I'm not qualified to review this, and I'm grateful for the work you've put in to it!

FRidh commented:
Interesting, I was unaware that `site.PREFIXES` was actually used for something. The documentation on it is also appalling.

As I mentioned in https://github.com/NixOS/nixpkgs/pull/82405 I want to look this weekend in creating a function in `site.py` that works similar to `venv` but that we can use. I intend to intrroduce an env `NIX_PYTHONHOME` that would replaces the porposed `NIX_PYTHONPREFIX`, `NIX_PYTHONEXECUTABLE` and the main use case of `NIX_PYTHONPATH`.

FRidh commented:
Also, we really need test cases for these things in Nixpkgs.

FRidh commented:
See `test_prefix` in https://github.com/NixOS/nixpkgs/pull/82560. This is actually the same that caused https://github.com/NixOS/nixpkgs/issues/66366.

FRidh commented:
For now, let's go ahead with this. The Python tests (https://github.com/NixOS/nixpkgs/pull/82560) are also in staging, it would be good if we update those to consider the effects of this PR.

FRidh committed:
Fix sys.prefix in case of a Nix env

The prefix will now be correct in case of Nix env.

Note, however, that creating a venv from a Nix env still does not function. This does not seem to be possible
with the current approach either, because venv will copy or symlink our Python wrapper. In case it symlinks
(the default) it won't see a pyvenv.cfg. If it is copied I think it should function but it does not...

adisbladis committed:
Python Add test for NIX_PYTHONPREFIX

FRidh commented:
Fixed `sys.prefix` with https://github.com/FRidh/nixpkgs/commit/1329fecaadd44d43ef2b3e14f3452df01a933738 on this branch.

adisbladis committed:
Python: Add integration test verifying NIX_PYTHONPATH with Mypy

adisbladis commented:
@FRidh I think this is ready for review now. I'm particularly interested in your opinion on `integrationTests`.

FRidh commented on pkgs/development/interpreters/python/tests/test_python.py:
```diff
@@ -35,6 +35,9 @@ def test_prefix(self):
         self.assertTrue(sys.prefix in site.PREFIXES)
 
+        if IS_NIXENV:
+            self.assertTrue(ENV in site.PREFIXES)
```
I think this is already handled. In `test_prefix` we already check if `sys.prefix == ENV`.

adisbladis replied:
You're right. I've removed this bit.

FRidh commented on pkgs/development/interpreters/python/tests.nix:
```diff
@@ -50,6 +51,14 @@ let
 
+  integrationTests = lib.recurseIntoAttrs {
+    # Before the addition of NIX_PYTHONPREFIX mypy was broken
+    nix-pythonprefix-mypy = import ./tests/test_nix_pythonprefix {
```
Every interpreter has this set of tests. Therefore, don't use `pkgs.python3`. If it only works with `python3`, use something like `python.isPy3k` to restrict.

adisbladis replied:
Done.

FRidh commented on pkgs/development/interpreters/python/tests/test_nix_pythonprefix/default.nix:
```diff
@@ -0,0 +1,25 @@
+{ interpreter, pkgs }:
```
more specific parameters than `pkgs`?

adisbladis replied:
Agreed, also switched to using `callPackage`.

FRidh commented on pkgs/development/interpreters/python/tests/test_nix_pythonprefix/typeddep/default.nix:
```diff
@@ -0,0 +1,11 @@
+  pname = "typeddep";
+  version = "1.3.3.7";
+
+  src = ./.;
```
not very important but this will include this nix file.

adisbladis replied:
I figured it didn't really matter and I didn't want any source filtering obscuring anything.

FRidh commented on testproject/default.nix:
```diff
@@ -0,0 +1,25 @@
+{ pkgs ? import <nixpkgs> {} }:
```
these testproject files need to be removed.

adisbladis replied:
Doh! I thought I did but probably got confused by a rebase.

FRidh reviewed (changes requested):
I think it's good we add these type of tests. Note it needs some fixing because now it will always execute with `python3`, even when testing e.g. `pypy`.

adisbladis commented:
@FRidh If you're happy with this PR could you please approve it without actually merging it. I'd like to make sure pypy tests are passing locally first but it takes ages to build.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-use-mypy-with-nixpkgs-cannot-find-implementation-or-library-stub-for-module/6639/2

iwinux commented:
It seems to breaks my previous working setup of `pip install --user`.
